### PR TITLE
media: i2c: ov5647: Support HFLIP and VFLIP

### PR DESCRIPTION
**Pull request by davidplowman**
Add these missing V4L2 controls. Tested binned and full resolution
modes in all four orientations using Raspberry Pi running libcamera.

Signed-off-by: David Plowman <david.plowman@raspberrypi.com>

### CODE REVIEW

**Comment by davidplowman** [comment]
@naushir @6by9 Well, this lot seems to fix the problems, but not sure if it's the "best" way...

File: drivers/media/i2c/ov5647.c [hunk 1]
```diff
@@ -152,7 +156,7 @@ static struct regval_list ov5647_640x480_8bit[] = {
 	{0x3036, 0x46},
 	{0x303c, 0x11},
 	{0x3106, 0xf5},
-	{0x3821, 0x07},
+	{0x3821, 0x03},
```
Comment (6by9) [inline drivers/media/i2c/ov5647.c]: I think my only comment is that we could drop bit 1 from these registers totally, as the control handler will always be called via __v4l2_ctrl_handler_setup and set it as defined by the control. That only leaves the binning bit defined.

Comment (davidplowman) [reply drivers/media/i2c/ov5647.c]: My only thought was for the register set on its own to correspond to "no flips at all", which, bizarrely, involves setting the hflip bit to counteract the secret internal flip. But given that we don't document anything about the registers anyway - I don't suppose it makes any difference really!

**Comment by davidplowman** [comment]
New version with those bit 1 values cleared - do we prefer this?

**Comment by 6by9** [comment]
I think I'm happy with that. It just seemed odd that we had bits in the registers that were never going to be used as set.

File: drivers/media/i2c/ov5647.c [hunk 2]
```diff
@@ -1140,14 +1171,16 @@ static int ov5647_enum_frame_size(struct v4l2_subdev *sd,
 {
 	struct ov5647_mode *mode = NULL;
 
-	if (fse->code == MEDIA_BUS_FMT_SBGGR8_1X8) {
+	if (fse->code == ov5647_get_mbus_code(sd, 1)) {
 		if (fse->index >= ARRAY_SIZE(supported_modes_8bit))
 			return -EINVAL;
 		mode = &supported_modes_8bit[fse->index];
-	} else if (fse->code == MEDIA_BUS_FMT_SBGGR10_1X10) {
+		fse->code = ov5647_get_mbus_code(sd, 1);
```
Comment (pelwell) [inline drivers/media/i2c/ov5647.c]: Is this not a no-op?
```
    if (a == func()) {
        a = func();
```
looks wrong.

Comment (davidplowman) [reply drivers/media/i2c/ov5647.c]: Yes actually, that would seem to be pointless! Let me fix it up...

**Comment by pelwell** [comment]
This patch does not apply cleanly to rpi-5.15.y - can someone more au-fait with V4L2 take a look?

**Comment by davidplowman** [comment]
I had a quick look at this. It seems to me like the 5.15 version simply doesn't support 8 bit modes at all, whereas our 5.10 one has had a 640x480 8-bit mode for some time. I'm thinking therefore that the 5.15 version is basically old and needs to be completely overwritten by the new one, but would appreciate Dave's opinion on that.

**Comment by 6by9** [comment]
```
commit 38c22308181562d52926b6701327c3fe209f6d84
Author: Jacopo Mondi <jacopo@jmondi.org>
Date:   Thu Nov 19 17:37:53 2020 +0100

    media: ov5647: Remove 640x480 SBGGR8 mode
    
    Capturing in 640x480 SBGGR8_1X8 hangs the system when capturing
    with the unicam driver on RaspberryPi 4 platform.
    
    Remove it and remove the support for multiple media bus codes in
    the driver.
    
    Signed-off-by: Jacopo Mondi <jacopo@jmondi.org>
    Signed-off-by: Sakari Ailus <sakari.ailus@linux.intel.com>
    Signed-off-by: Mauro Carvalho Chehab <mchehab+huawei@kernel.org>
```
So upstream have dropped the 8bit modes. That should simplify things somewhat.

**Comment by davidplowman** [comment]
I don't think I like that. Our users explicitly want to be able to get the 8-bit mode because it frames faster.

**Comment by 6by9** [comment]
Worth testing whether it actually works then as the commit text implies it doesn't.
IIRC it was also a mess in that it was the top left quarter of the image because it was binned and then cropped, not binned and skipped.

**Comment by davidplowman** [comment]
Ah yes, I remember that. Maybe that was the broken mode. Perhaps I'm remembering a different sensor where they wanted a mode with a different bit-depth. Maybe the imx477. In which case this I guess this is OK.

**Pull request merged**